keps/sig-auth/541-external-credential-providers/README.md

```diff
@@ -127,24 +157,49 @@ be readable and executable by the client process.
 `args` specifies extra arguments passed to the executable.
 
 `env` specifies environment variables to pass to the provider. The environment
-variables set in the client process are not passed.
+variables set in the client process are also passed to the provider.
 
 ### Provider input format
 
-```
+```json
 {
-  "apiVersion": "client.authentication.k8s.io/<version>",
-  "kind": "ExecCredential"
+  "apiVersion": "client.authentication.k8s.io/v1beta1",
+  "kind": "ExecCredential",
+  "spec": {
+    "cluster": {
+      "server": "https://1.2.3.4:8080",
+      "certificate-authority": "/etc/kubernetes/ca.pem"
+    }
+  }
 }
 ```
 
-Provider can safely ignore `stdin` since input object doesn't carry any data.
+The `spec.cluster` field is the current cluster that the client is communicating
+with (i.e it is the cluster the client knows it must communicate with after it
+has completed parsing its `kubeconfig`, flags and environment variables).
+This struct is defined in [k8s.io/client-go/tools/clientcmd/api/v1#Cluster](https://pkg.go.dev/k8s.io/client-go/tools/clientcmd/api/v1?tab=doc#Cluster)
```

Comment:
Depending on clientcmd here might not be ideal given this is one of the oldest API groups and has a lot of problems. One alternative is that we surface useful info in a type surfaced in this package. Questions:
* How does this work with clients create from rest.Config?
* What are the useful fields in Cluster that we want to forward?

Reply:
> Depending on clientcmd here might not be ideal given this is one of the oldest API groups and has a lot of problems.

The `ExecConfig` API itself is located in `clientcmd` (though I understand the desire not to use a `struct` from the earliest days of Kube).

> One alternative is that we surface useful info in a type surfaced in this package.

It ends up being quite similar to the `Cluster` type (see below).

> How does this work with clients create from rest.Config?

We have to expand `rest.Config` is some way to carry the `[]NamedExtension` information through.  The rest of the information is already present.

> What are the useful fields in Cluster that we want to forward? NamedExtensions? 

All of them.  The hostname and TLS config allows communication with the cluster (ex: to get some dynamic config from the cluster itself possibly by reading some well-known location).  Having a `[]NamedExtensions` vs. a single `runtime.RawExtension` is not meaningfully different since you can still pass arbitrary config via either mechanism.

```go
// Cluster contains information about how to communicate with a kubernetes cluster
type Cluster struct {
	// Server is the address of the kubernetes cluster (https://hostname:port).
	Server string `json:"server"`
	// TLSServerName is used to check server certificate. If TLSServerName is empty, the hostname used to contact the server is used.
	// +optional
	TLSServerName string `json:"tls-server-name,omitempty"`
	// InsecureSkipTLSVerify skips the validity check for the server's certificate. This will make your HTTPS connections insecure.
	// +optional
	InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
	// CertificateAuthority is the path to a cert file for the certificate authority.
	// +optional
	CertificateAuthority string `json:"certificate-authority,omitempty"`
	// CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
	// +optional
	CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
	// Extensions holds additional information. This is useful for extenders so that reads and writes don't clobber unknown fields
	// +optional
	Extensions []NamedExtension `json:"extensions,omitempty"`
}
```

If I was to create a new `struct` it may look like:

```go
// ExecCluster contains information to allow an exec plugin to communicate with the kubernetes cluster being authenticated to.
type ExecCluster struct {
	// Server is the address of the kubernetes cluster (https://hostname:port).
	Server string `json:"server"`
	// TLSServerName is used to check server certificate. If TLSServerName is empty, the hostname used to contact the server is used.
	// +optional
	TLSServerName string `json:"tls-server-name,omitempty"`
	// CertificateAuthorityData contains PEM-encoded certificate authority certificates.
	// If empty, system roots are used.
	// +optional
	CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
	// Config holds additional config data that is specific to the exec plugin with regards to the cluster being authenticated to.
	// +optional
	Config []NamedExtension `json:"config,omitempty"`
}
```

> Notably "name" of the Cluster is not in the Cluster object itself.

Trying to wire in the `name` caused a lot of ugly code and I could not think of why it would be needed.

---

I do like the `ExecCluster` struct better since it is cleaner and streamlined.  Since the user needs to be able to specify a per-cluster config in the cluster object itself, we must use the `[]NamedExtension` from the `Cluster` object as input (I do not see us expanding or modifying the `Cluster` `struct`).

Thoughts?

Reply:
To help clarify, here is what a `kubeconfig` may look like with a shared `exec` config and distinct configuration per cluster:

```yaml
apiVersion: v1
kind: Config

users:
- name: user-for-all-clusters
  user:
    exec:
      apiVersion: "client.authentication.k8s.io/v1beta1"
      command: "/bin/exec-plugin"
      env:
      - name: "FOO"
        value: "some env var that is always the same"
      args:
      - "some arg that is always the same"

clusters:
- name: cluster1
  cluster:
    server: "https://cluster1"
    extensions:
    - name: exec
      extension:
        audience: audience1  # <--- custom config for cluster1
- name: cluster2
  cluster:
    server: "https://cluster2"
    extensions:
    - name: exec
      extension:
        audience: audience2  # <--- custom config for cluster2

contexts:
- name: cluster1
  context:
    cluster: cluster1
    user: user-for-all-clusters  # <--- always the same
- name: cluster2
  context:
    cluster: cluster2
    user: user-for-all-clusters  # <--- always the same

current-context: cluster1
```

Note the distinct `audience` information that is used per cluster.  If we wanted to formalize onto a single key from the `extensions` list, we could reserve the key `exec` for this purpose.

```yaml
    extensions:
    - name: exec    # <--- this is the reserved key
      extension:    # <--- this is the config for the exec plugin
        audience: audience1
```

Then we could have:

```go
type ExecCluster struct {
	// ... same as above

	// Config holds additional config data that is specific to the exec plugin with regards to the cluster being authenticated to.
	// +optional
	Config runtime.RawExtension `json:"config,omitempty"`
}
```

Reply:
Thanks for the detailed reply.I don't feel terribly strongly either way, but I currently lean towards just copying the struct as well. The cost looks minimal right now (given we care about 4 fields) and copying could save us some future pain.